Use partition/rpartition instead of split/rsplit

akx opened the pull request:
## Proposed change

I spotted a couple of places (and then a couple more . . .) where uses of `.split` and `.partition` could be replaced with the more efficient but [not particularly well-known `partition`](https://docs.python.org/3.10/library/stdtypes.html#str.partition) function (and same for `rsplit`/`rpartition`).

* partition/rpartition always returns a 3-tuple so checks for how many pieces are returned aren't necessary anymore. This simplifies a bunch of code (a bit).
* It's also approximately 20% faster than split in a simple case (and yes, it's a microbenchmark but every little counts in my books); see timeit comparison below
* constructing tuples is, to my best knowledge, more memory-efficient than lists, so a modest memory usage improvement could also be in the cards.

I understand that this is a bit on the big side as far as the number of touched files/components goes, but it should nevertheless be easy to review:

* The changes are, in most cases, a single line per file. (Looks like opening this triggered quite some codeowners for review...)
  * In some places, I've pulled a value that's being `.split()` twice (or similar) into a local variable, for performance and readability.
* The changes are split into three commits based on the changed idioms.

If you like, I have no problem splitting this into smaller PRs either (though I'd like some guidance on what the scope of each PR should then be).

## Benchmarking getting the first or last bit of a delimited string

```
$ python timeit-compare.py -n 10000000 -s 's = "a.b"' 's.partition(".")[0]' 's.partition(".")[-1]' 's.split(".", 1)[0]' 's.split(".", 1)[-1]' 's.rpartition(".r")[0]' 's.rpartition(".")[-1]' 's.rsplit(".", 1)[0]' 's.rsplit(".", 1)[-1]'
s.partition(".")[0]   : 10000000 loop(s), best of 5: 0.079 usec per loop
s.partition(".")[-1]  : 10000000 loop(s), best of 5: 0.081 usec per loop
s.split(".", 1)[0]    : 10000000 loop(s), best of 5: 0.094 usec per loop
s.split(".", 1)[-1]   : 10000000 loop(s), best of 5: 0.092 usec per loop
s.rpartition(".r")[0] : 10000000 loop(s), best of 5: 0.071 usec per loop
s.rpartition(".")[-1] : 10000000 loop(s), best of 5: 0.084 usec per loop
s.rsplit(".", 1)[0]   : 10000000 loop(s), best of 5: 0.091 usec per loop
s.rsplit(".", 1)[-1]  : 10000000 loop(s), best of 5: 0.095 usec per loop
```


## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Checklist

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.
  - No new tests; behavior should not have changed, and current tests should cover all of this (hopefully...) 

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

akx committed:
Replace .split/.rsplit(maxsplit=1) with partition/rpartition where feasible

akx committed:
Replace .split()[0] with .partition()[0]

akx committed:
Replace .split()[-1] with .rpartition()[-1]

bdraco commented:
While I'm not usually one to turn performance improvements, this PR changes quite files that don't have test coverage in places that are not called frequently. We don't want the risk of a regression from untested code without a tangible performance improvement in a hot codepath.

We could accept a change to areas you have identified as called frequently, but based on a quick profiler run that doesn't seem to be any of the ones in this PR.
<img width="466" alt="Screenshot 2022-11-03 at 12 31 29 PM" src="https://user-images.githubusercontent.com/663432/199797281-a43a47df-0dcb-46a9-9452-2b3518d66219.png">

akx commented:
@bdraco Thanks for the comments!

> While I'm not usually one to turn down performance improvements, this PR changes quite files that don't have test coverage in places that are not called frequently.

Fair. How can you tell which files changed here are or are not covered, as Codecov is having a tough time with this repo? (I.e. is there a way I could tell too?)

> We don't want the risk of a regression from untested code without a tangible performance improvement in a hot codepath.

I understand, but on the other hand if the same `.split(...)[...]` pattern works in one place, why wouldn't it work elsewhere?

> based on a quick profiler run that doesn't seem to be any of the ones in this PR.

Can you share the methodology for that?

bdraco commented:
> @bdraco Thanks for the comments!
> 
> > While I'm not usually one to turn down performance improvements, this PR changes quite files that don't have test coverage in places that are not called frequently.
> 
> Fair. How can you tell which files changed here are or are not covered, as Codecov is having a tough time with this repo? (I.e. is there a way I could tell too?)
>

You can manually check a each integration with:
`pytest --cov=homeassistant/components/wiz/ --cov-report term-missing -- tests/components/wiz/`


> > We don't want the risk of a regression from untested code without a tangible performance improvement in a hot codepath.
> 
> I understand, but on the other hand if the same `.split(...)[...]` pattern works in one place, why wouldn't it work elsewhere?

The usage is not the exact same everywhere and there is always a risk of regression with making any change.  If the code isn't tested with test coverage it needs to be tested manually before we can accept changes.

> 
> > based on a quick profiler run that doesn't seem to be any of the ones in this PR.
> 
> Can you share the methodology for that?

Run the profiler integration for 60s https://www.home-assistant.io/integrations/profiler/#service-profilerstart

Its going to generate a cProfile and convert it to a callgrind.out.xxx file for you which can be opened with qcachegrind

akx commented:
@bdraco

Thanks for the pointers, I appreciate it.

I rebased this PR and removed changes to files that match any `.coveragerc` omit stanza (see full list at https://gist.github.com/akx/ffa23728c32d76f800b7a8218216bc08). That should leave only changes to files that should be covered by existing tests.

bdraco commented:
> @bdraco
> 
> Thanks for the pointers, I appreciate it.
> 
> I rebased this PR and removed changes to files that match any `.coveragerc` omit stanza (see full list at [gist.github.com/akx/ffa23728c32d76f800b7a8218216bc08](https://gist.github.com/akx/ffa23728c32d76f800b7a8218216bc08)). That should leave only changes to files that should be covered by existing tests.

Thats is an improvement, but unfortunately just because its not in `.coveragerc` doesn't mean the line is covered as the code doesn't have 100% coverage.  Additionally you've got at least 25 reviewers that this change affects who potentially need to review this PR so its better to break up a change like this into smaller PRs, instead I'd urge you to focus on the the changes that will have the highest impact. 

Sometimes that means only changing very few lines that are executed the most frequently as that change can have more impact and doesn't cost a whole lot of reviewer time.

Rudd-O commented on homeassistant/auth/providers/command_line.py:
```diff
@@ -90,7 +90,7 @@ async def async_validate_login(self, username: str, password: str) -> None:
                         continue
-                    key, value = line.split("=", 1)
+                    key, _, value = line.partition("=")
                 except ValueError:
```
This won't work because ValueError is intended to catch the partition error.  Better leave as is or handle an empty separator (item 1 in returned tuple) by raising ValueError..

Rudd-O commented on homeassistant/components/command_line/sensor.py:
```diff
@@ -152,7 +152,7 @@ def update(self) -> None:
             args = None
             args_compiled = None
         else:
-            prog, args = command.split(" ", 1)
```
This should probably not be done this way (**security issue**)  Please use `shlex.split()` and take the first argument of the returned list for prog, the rest of the list would be args.

Good catch.

akx replied:
Not a security issue -- you are supposed to be able to execute arbitrary shell commands anyway. See comments: https://github.com/akx/home-assistant-core/blob/cc41c6d10bce6c7ad0afa5f7cdba69c354918102/homeassistant/components/command_line/__init__.py#L44

akx replied:
The `args` aren't passed directly on though, but via a `Template`. That could make a difference here. Either way, this PR is likely not going to go anywhere as-is due to too large a scope, so these concerns should be done in the (possible) future per-component PRs.

Rudd-O commented on homeassistant/components/filesize/config_flow.py:
```diff
@@ -58,7 +58,7 @@ async def async_step_user(
                 self._abort_if_unique_id_configured()
 
-                name = str(user_input[CONF_FILE_PATH]).rsplit("/", maxsplit=1)[-1]
+                name = str(user_input[CONF_FILE_PATH]).rpartition("/")[-1]
```
This should probably not be done this way.  Use `os.path.basename()`.

Rudd-O commented on homeassistant/components/image/__init__.py:
```diff
@@ -205,8 +205,13 @@ async def get(self, request: web.Request, image_id: str, filename: str):
     image = ImageOps.exif_transpose(Image.open(original_path))
     image.thumbnail(target_size)
-    image.save(target_path, format=content_type.split("/", 1)[1])
+    image.save(target_path, format=content_type.partition("/")[-1])
```
Note that content-type can also include `; blah` tacked on at the end, so recommending another partition to remove the tail after the semicolon, followed by an rstrip.  Other code similar to what I'm asking looks like this: `content_type.partition(";")[0]`.

Rudd-O commented on homeassistant/components/lovelace/cast.py:
```diff
@@ -126,7 +126,7 @@ async def async_play_media(
 
     if "/" in media_id:
-        url_path, view_path = media_id.split("/", 1)
+        url_path, _, view_path = media_id.partition("/")
```
Double check that url_path is the one meant to be without slashes.  (Yes, the new code has no functional change and is equivalent to the prior code.)

Rudd-O commented on homeassistant/components/media_source/local_source.py:
```diff
@@ -152,7 +152,7 @@ def _build_item_response(
         # Check that it's a media file
         if is_file and (
-            not mime_type or mime_type.split("/")[0] not in MEDIA_MIME_TYPES
+            not mime_type or mime_type.partition("/")[0] not in MEDIA_MIME_TYPES
```
Same deal, perhaps the rest of the MIME type needs to `content_type.partition(";")[0]`.

Rudd-O commented on homeassistant/components/plex/server.py:
```diff
@@ -187,7 +187,7 @@ def _update_plexdirect_hostname():
                     error = error.__context__
                 if isinstance(error, ssl.SSLCertVerificationError):
-                    domain = urlparse(self._url).netloc.split(":")[0]
+                    domain = urlparse(self._url).netloc.partition(":")[0]
```
How did this ever work?

```
>>> urlparse("https://user:password@google.com/abcdef")
ParseResult(scheme='https', netloc='user:password@google.com', path='/abcdef', params='', query='', fragment='')
```

What is desired at least according to spec is to split at the @ spot.  Does the Plex integration do something funny with URLs that would cause `urlparse` to give back a colon but no at sign?

akx replied:
A Plex URL would never have basic auth parameters, but it could have a port (`myplexserver:1234`).

Rudd-O replied:
OK.

Rudd-O commented on homeassistant/components/vizio/config_flow.py:
```diff
@@ -98,9 +98,9 @@ def _get_pairing_schema(input_dict: dict[str, Any] | None = None) -> vol.Schema:
 def _host_is_same(host1: str, host2: str) -> bool:
     """Check if host1 and host2 are the same."""
-    host1 = host1.split(":")[0]
+    host1 = host1.partition(":")[0]
```
I am not sure how this comparison worked, since host1 and host2 can be IPv6 addresses and separating at the colon will just yield something like `00`.

This needs further investigation.  Meanwhile, this can be used to switch on the case that the input values are NOT IP addresses:

```
>>> ip_address("roxanne.dragonfear")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python3.10/ipaddress.py", line 54, in ip_address
    raise ValueError(f'{address!r} does not appear to be an IPv4 or IPv6 address')
ValueError: 'roxanne.dragonfear' does not appear to be an IPv4 or IPv6 address
```

akx replied:
My guess is that Vizio equipment doesn't do IPv6, so that's a non-issue. 😁

Rudd-O replied:
Let's postulate that.

Will we remember this issue in the function down the road 3 or 4 years from now, when Vizio TVs begin supporting IPv6?

akx commented on homeassistant/components/here_travel_time/sensor.py:
```diff
@@ -160,9 +160,10 @@ def __init__(
     def extra_state_attributes(self) -> Mapping[str, Any] | None:
         """GPS coordinates."""
         if self.coordinator.data is not None:
+            lat, _, lon = self.coordinator.data[ATTR_ORIGIN].partition(",")
```
That's irrelevant to this PR - this would only touch the splits.

Rudd-O replied:
I wouldn't mind if this specific code remained functionally the same as the partition change takes place.  However, please look at the _other_ comments — there are definitely faulty assumptions or even potential security issues in the old code.

bdraco commented:
Before you spend any more time on this, I want to point out this PR is not likely to get merged in its current form.

You'll have much better luck with smaller PRs targeting specific integrations so the codeowners of that integration have a chance to review it instead of everyone at once. (please don't go open 100s of PRs for this)

https://developers.home-assistant.io/docs/creating_component_code_review#5-make-your-pull-request-as-small-as-possible

balloob commented:
The change is a good one, but like bdraco said, let's do smaller PRs.
